web3.js/src/connection.ts

```diff
@@ -2162,14 +2160,26 @@ export type ConfirmedSignatureInfo = {
  */
 export type HttpHeaders = {[header: string]: string};
 
+export type AlterReturnType<TFn, TReturn> = TFn extends (
+  ...args: infer A
+) => any
+  ? (...args: A) => TReturn
+  : never;
+export type AddParameters<
+  TFunction extends (...args: any) => any,
+  TParameters extends [...args: any],
+> = (
+  ...args: [...Parameters<TFunction>, ...TParameters]
+) => ReturnType<TFunction>;
+export type FetchFn = typeof globalThis.fetch;
+
 /**
  * A callback used to augment the outgoing HTTP request
  */
-export type FetchMiddleware = (
-  url: string,
-  options: any,
-  fetch: (modifiedUrl: string, modifiedOptions: any) => void,
-) => void;
+export type FetchMiddleware = AlterReturnType<
+  AddParameters<FetchFn, [fetch: AlterReturnType<FetchFn, void>]>,
+  void
+>;
```

Comment:
What is the benefit of expressing the type of `FetchMiddleware` like this over the original type signature which is more explicit?

Reply:
Well, the original signature is just straight up wrong. The first parameter to `fetch` is `RequestInfo` not `string`. The second `RequestInit` not `any`. I'm just trying to make everything properly line up with the type of the `fetch-impl` shim.